### PR TITLE
Branches-Sara

### PR DESCRIPTION
**Pull request by sarashahbaig**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | A model is like a class; it talks to the database, stores and validates data.
Describe in your own words what the Controller is doing in Rails | Controller dos the work of analyzing user requests, data submissions, cookies and sessions.  
Describe in your own words what the View is doing in Rails | Views are what the user sees as HTML, CSS.
Describe an edge-case controller test you wrote |get edit_task_path(-1) must_respond_with :redirect.
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params helps protect against assignment to the model and gives us control to do so in the controller where it belongs.
How are Rails migrations related to Rails models?|Requirements change all the time, and those changes often lead to database changes which are in Rails models.
Describe one area of Rails that are still unclear on | Nothing in particular.

### CODE REVIEW

**Comment by tildeee** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | x
Answered comprehension questions | x
Successfully handles: Index, Show | x
Index & Show tests pass | x
Successfully handles: New, Create | x
New & Create tests pass | x
Successfully handles: Edit, Update | x
Tests for Edit & Update test valid & invalid task ids | x
Successfully handles: Destroy, Task Complete | x
Tests for Destroy & Task Complete include tests for valid and invalid task ids | x, missing edge cases
Routes follow RESTful conventions | could be improved
Uses named routes (like `_path`) | x
**Overall** |



Well done on this project, Sara! Your code generally looks good, and your tests generally look good.


However, you have a couple of places that I have some concern about for syntax, with possibly some concerns for understanding. I hope that when you read these comments, you'll have a better sense for ways to improve syntax, areas of concept to review, and questions for me or for any instructor.
- routes
- edge cases on tests for controllers
- updating models using dot syntax (`.`)
- using `find_by` to determine if a Task is missing




Let me know what questions you have! Keep up the hard work!

File: app/assets/stylesheets/application.css [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+section h1 {
+  font-size: 25px;
+  text-transform: uppercase;
+}
+a {
+  font-size: 12px;
+  text-decoration: none;
+  color: white;
```
Comment (tildeee) [inline app/assets/stylesheets/application.css:41]: Because of this style, which makes all links turn white, it was impossible for me to see the link to Mark Complete or Mark Incomplete, so I almost thought that this feature wasn't done! Watch out with your styles and think about what the user sees.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -106,21 +114,86 @@
...
 
+        expect {
+          delete task_delete_path(id)
+        }.must_change "Task.count", 1
+
+        must_respond_with :redirect
+        must_redirect_to root_path
+
+        expect( Task.find_by(id: id) ).must_be_nil
+    end
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:173]: What about a test for deleting a task that doesn't exist?

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -106,21 +114,86 @@
...
+          
+      update_task_from_data = {
+        task:{
+          name: "Watch TV",
+          description: "Descripton for tv",
+          completed: "yes"
+        }
+      }
+      #Act
+      patch task_path(existing_task.id, params: update_task_from_data)
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:190]: This test doesn't use your `complete_task_path` route, and uses a completely different route/action here. In fact, it doesn't even go into the `TasksController` `toggle_complete` action here!

File: test/controllers/tasks_controller_test.rb [hunk 4]
```diff
@@ -106,21 +114,86 @@
...
+      must_redirect_to task_path
+      
+      #Assert
+      expect( Task.find_by(id: existing_task.id).completed).wont_equal nil
+    end
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:195]: What happens if we try to `toggle_complete` on a task that doesn't exist? It would be awesome to have a test here on that.

File: config/routes.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+#   post '/tasks', to: 'tasks#create'
+
+# # Routes that operate on individual tasks
+#   get '/tasks/:id', to: 'tasks#show', as: "task"
+
+#   get '/tasks/:id/edit', to: 'tasks#edit', as: 'edit_task'
+
+  #  patch '/tasks/update/:id', to: 'tasks#update', as: 'task_update'
+
+  delete '/tasks/:id', to: 'tasks#destroy', as: 'task_delete'
```
Comment (tildeee) [inline config/routes.rb:23]: Hm, here you make a new route to delete a task. This is a route that can be generated with the `resources`, like the in line above with `resources :tasks`

File: config/routes.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+#   get '/tasks/new', to: 'tasks#new', as: 'new_task' 
+# # Step 2: Create an action so that the form data actually gets processed by Rails (by the server) and creates that new task, and changes the database
+#   post '/tasks', to: 'tasks#create'
+
+# # Routes that operate on individual tasks
+#   get '/tasks/:id', to: 'tasks#show', as: "task"
+
+#   get '/tasks/:id/edit', to: 'tasks#edit', as: 'edit_task'
+
+  #  patch '/tasks/update/:id', to: 'tasks#update', as: 'task_update'
```
Comment (tildeee) [inline config/routes.rb:21]: This file may be easier to work with if you delete the comments

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  delete '/tasks/:id', to: 'tasks#destroy', as: 'task_delete'
+
+  get '/tasks/complete/:id', to: 'tasks#toggle_complete', as: 'complete_task'
```
Comment (tildeee) [inline config/routes.rb:25]: What is the controller action `tasks#toggle_complete` (TasksController, toggle_complete action) supposed to do? It's supposed to change a task's data so that it's complete. Since it's going to change task data, instead of ask/query for data, we should think about the HTTP Verb for this route. The HTTP Verb you defined is `get`, which is usually associated with asking/querying for information. Since we're not querying, we should use the HTTP Verbs associated with changing information, like `patch` or `put`. Therefore, maybe the better version of this route may be `patch '/tasks/complete/:id', to: 'tasks#toggle_complete', as: 'complete_task'`

Of course, this will affect the link in the view that goes to this route.

File: app/controllers/tasks_controller.rb [hunk 8]
```diff
@@ -0,0 +1,117 @@
+class TasksController < ApplicationController
+    def index
+        @tasks = Task.all.order(:id)
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:3]: Nice detail to order and sort them!

File: app/controllers/tasks_controller.rb [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    end
+
+    def show
+        task_id = params[:id].to_i
+        @task = Task.find_by(id: task_id)
+
+        if task_id < 0
+            flash[:Error] = "Could not find task with id: -1"
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:11]: We haven't learned what `flash` is yet -- what are you doing here? :)

File: app/controllers/tasks_controller.rb [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        if id < 0
+            redirect_to root_path
+        end
+        @task = Task.find_by(id: id )
+        
+    end
+        
+    def update
+        id = params[:id].to_i
+        if id < 0
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:66]: Your conditional `if id < 0` is to help determine if the task was not found. However, this is not a very solid way of approaching the problem. This logic works if `params[:id]` is equal to `-1`. What if `params[:id]` is equal to `tacocat`? Then `id` would equal `0`. Also, what if `params[:id]` is equal to `999`, which is a valid number, but there isn't a task with that ID number?

The better thing to do would be to check if `Task.find_by` gives back a task or if it gives back `nil`.

Let me know if you have questions, because this is really important

File: app/controllers/tasks_controller.rb [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            redirect_to root_path
+        end
+        @task = Task.find_by(id: id)
+        @task[:name] = params[:task][:name]
+        @task[:description] = params[:task][:description]
+        @task[:completed] = params[:task][:completed]
+                
+        if @task.save
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:74]: Instead of re-assigning each part of `@task`, it will be better practice to use strong params, so to change this method to

```ruby
@task = Task.find_by(id: id)
if @task.update(task_params)
```

File: app/controllers/tasks_controller.rb [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            return
+        end
+    end
+    
+
+    def toggle_complete 
+        id = params[:id]
+        @task = Task.find_by(id: id)
+        if @task.completed == nil
+            @task[:completed] = "yes"
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:98]: Instead of using this syntax with the square brackets, the same thing can be accomplished with `.` and it is more Rails-y

```ruby
@task.completed = "yes"
```
